### PR TITLE
Update to wgpu 0.6.0

### PR DESCRIPTION
**Pull request by DevOrc**
Updates to wgpu-core 0.6.0. This is very much a WIP and many things still need to be implemented (textures, push constants), but I thought I would throw this up for comments. Also for error handling, anytime a result is an error, it just panics with a description of the function that panicked. 

 **Testing:** 
Triangle/Compute Examples

After I finish updating everything I will test it against my own stuff

### CODE REVIEW

**Comment by kvark** [comment]
Should this be marked as a "draft" until you finish?

**Comment by DevOrc** [comment]
oops! meant to do that 😄

**Comment by DevOrc** [comment]
This is hopefully finished now!

**Notes:**
- Updates wgpu-core to master branch ([rev 8ce2530b](https://github.com/gfx-rs/wgpu/tree/8ce2530b6963982ef9a58f7efac682d13ce4c508))
- Adds repr(C) versions for BufferCopyView, TextureCopyView, TextureViewDescriptor, CompareFunction, PipelineLayoutDescriptor, VertexBufferDescriptor, VertexStateDescriptor, RenderPipelineDescriptor, ComputePipelineDescriptor
- Adds error messages when wgpu-core returns and Err(_). 

**Testing**
wgpu-native and kgpu examples work

**cbingen bug note:**

The following pattern causes cbindgen to crash:
```rust
//In wgt
mod wgt {
    pub struct Foo<T> {
        foo: T,
    }
}

//In wgc
pub type Foo = wgt::Foo<u32>;

//In wgpu-native
mod native {
    #[repr(C)]
    pub struct Foo{
        foo: u32
    }
}
```
Basically, cbindgen will panic because it ignores the module name and therefore will have multiple types to export with the same name. This pattern is used by BufferCopyView and TextureCopyView. To work around it, the native versions of the struct are called BufferCopyViewC and TextureCopyViewC, and they are renamed when CBindgen generates the header. A solution would be to move TextureCopyView and BufferCopyView into wgpu-core and remove the type alias.

See https://github.com/eqrion/cbindgen/issues/573 for more details

**Comment by kvark** [comment]
> To work around it, the native versions of the struct are called BufferCopyViewC and TextureCopyViewC

I think we used the `Raw` suffix for those previously. I don't mind either way, let's just try to be consistent here.

File: cbindgen.toml [hunk 1]
```diff
@@ -27,7 +28,8 @@ style = "both"
 [export]
 prefix = "WGPU"
 include = [
-    "AnisotropicSamplerDescriptorExt"
+    "AnisotropicSamplerDescriptorExt",
+	"SamplerBorderColor"
```
Comment (kvark) [inline cbindgen.toml:32]: tab detected :)

File: src/command.rs [hunk 2]
```diff
@@ -28,67 +41,100 @@ pub extern "C" fn wgpu_command_encoder_copy_buffer_to_buffer(
...
+}
+
+#[repr(C)]
+#[derive(Clone)]
+pub struct BufferCopyViewC {
+    pub layout: wgt::TextureDataLayout,
+    pub buffer: id::BufferId,
+}
+
+impl Into<wgc::command::BufferCopyView> for BufferCopyViewC {
```
Comment (kvark) [inline src/command.rs:54]: it would be simpler to just have this as `pub fn to_wgpu(&self) -> wgc::command::BufferCopyView`. since it would turn `xxx.cloned().into()` into `xxx.to_wgpu()`, which is much cleaner

File: src/command.rs [hunk 3]
```diff
@@ -28,67 +41,100 @@ pub extern "C" fn wgpu_command_encoder_copy_buffer_to_buffer(
         destination,
         destination_offset,
         size))
+    .expect("Unable to copy buffer to buffer")
+}
+
+#[repr(C)]
+#[derive(Clone)]
+pub struct BufferCopyViewC {
```
Comment (kvark) [inline src/command.rs:49]: would it help to slap `repr(C)` on the relevant `wgt` structs?

File: src/lib.rs [hunk 4]
```diff
@@ -23,8 +26,11 @@ impl OwnedLabel {
             )
         })
     }
-    fn as_ref(&self) -> Option<&str> {
-        self.0.as_ref().map(|s| s.as_str())
+    fn ref_cow(&self) -> Option<Cow<str>> {
+        self.0.as_ref().map(|s| Cow::Borrowed(s.as_str()))
+    }
+    fn owned_cow<'a>(self) -> Option<Cow<'a, str>> {
```
Comment (kvark) [inline src/lib.rs:32]: should be named `into_cow`

File: src/lib.rs [hunk 5]
```diff
@@ -23,8 +26,11 @@ impl OwnedLabel {
             )
         })
     }
-    fn as_ref(&self) -> Option<&str> {
-        self.0.as_ref().map(|s| s.as_str())
+    fn ref_cow(&self) -> Option<Cow<str>> {
```
Comment (kvark) [inline src/lib.rs:29]: we can call it `as_cow`

**Review by kvark — COMMENTED** [review]
Amazing work here, thank you! A few notes before we proceed

File: src/command.rs (hunk 3 above)
Comment (DevOrc) [reply src/command.rs:49]: We can do that for now but we will eventually need to make a separate version here to match the [WebGPU headers](https://github.com/webgpu-native/webgpu-headers/blob/c761a96381b73f48d8e0416d0b7b4767e10a3860/webgpu.h#L750)

Comment (kvark) [reply src/command.rs:49]: Right, so maybe we should try too hard to re-use it, I agree!

**Comment by DevOrc** [comment]
Ok all of those recommendations should be fixed!

**Review by kvark — APPROVED** [review]
bors r+
